README
Interf (alpha)
--------------

- as yet unreleased

 'Interf' is a program that generates images of interference patterns
 based on a number of point source co-ordinates. These images can then
 be combined using an external program (e.g. apngasm) to create an
 animation.

 Copyright - C Hancock (Chris2048)

Dependencies
------------

 * Python 2.7
  * Python Imaging Library (PIL)
  * NumPy
  * plac <plac.googlecode.com>
  * clint <github.com/kennethreitz/clint>

Recommended
-----------

 * apngasm - to assemble png files into apng.

Installation
------------

 Simply execute 'interf.py', use '-h' switch for options.

Contact
-------

 Post bugs, fixes, and other issues on github.com/Chris2048



ChangeLog
Wed, 02 Nov 2011 13:17:45 +0000 Chris2048

- as yet unreleased

Mon, 06 Aug 2012 21:50:25 +0000 Chris2048

- put project on Github



interf_test.py
#!/usr/bin/env python2.7

from random import random, randint
#import os
from pyvows import Vows, expect
import interf  # can this be implicit?


@Vows.create_assertions
def to_be_less_than(topic, expected):
    return topic < expected


class FooError(Exception):
    pass


@Vows.assertion
def not_to_be_an_error_like(topic, expected):
    if isinstance(topic, expected):
        raise Vows.VowsAssertionError('''Expected topic(%s) not to be
                an error of type %s, but it was a %s
                ''', topic, expected, topic.__class__)


# TODO create general 'value at point' test classes, for use in
# specific and random tests
@Vows.batch
class ValueAtPointTests(Vows.Context):
    def topic(self):
        point_test_data = (
            (1.0 / 2, 1.0 / 8, 1.0 / 4, 0.75),
            (1.0 / 3, 3.0 / 5, 1, 0.5),
            (1.0 / 8, 2.0 / 5, 1.0 / 3, 1), )
        base = interf.Base(100, 100, 100, None, None, None)
        for p in point_test_data:
            base.addPoint(*p)
        return base

    def check_some_values(self, topic):
        # manually pre-calculate some values? supply an error as well
        expect(len(topic.points)).to_equal(3)
        expect(interf.value_at_point(topic, 50, 50, 1, 0)).to_equal(
                9.166154761796539e-1)
        expect(interf.value_at_point(topic, 51, 50, 1, 0)).to_equal(
                9.538223520374552e-1)
        expect(interf.value_at_point(topic, 50, 51, 1, 0)).to_equal(
                8.966206758021245e-1)
        expect(interf.value_at_point(topic, 51, 51, 1, 0)).to_equal(
                9.345812279980938e-1)


@Vows.batch
class GeneratePicsFileTest(Vows.Context):
    def topic(self):
        err_file_test_data = [
            (None, AttributeError),
            ('/tmp/nonexistant-directorfoo/foo.foo', IOError), ]
        for td in err_file_test_data:
            excep = None
            try:
                base = interf.Base(100, 100, 100, 3000, td[0], '2')
                # will cause ZeroDivisionError otherwise
                base.addPoint(0, 0, 0, 0)
                interf.generate_pics(base)
            except Exception as ex:
                excep = ex
            yield (excep, td[1])

    def check_topic_error(self, topic):
        expect(topic[0]).to_be_an_error_like(topic[1])
        expect(topic[0]).Not.to_be_an_error_like(FooError)


@Vows.batch
class BaseTests(Vows.Context):
    def topic(self):
        base_test_data = [
            (0, 0, 100, None, None, None),
            (0, 0, 1000, None, None, None), ]
        for item in base_test_data:
            yield interf.Base(*item)

    def resolution_is_10(self, topic):
        if topic.ft == 100:
            expect(topic.resolution).to_equal(10)

    def not_different_if_floats_are_used(self, topic):
        expect(topic.resolution).to_equal(1000.0 / topic.ft)

# A reminder on how to check file contents

#@Vows.batch
#class ReadFile(Vows.Context):
#    def topic(self):
#        return os.path.exists('some_file.txt')
#
#    class AfterSuccessfullyReading(Vows.Context):
#        def topic(self, exists):
#            if exists:
#                return open('some_file.txt').read()
#            return None
#
#        def if_exists_we_can_read_the_contents(self, topic):
#            expect(topic).to_be_like('some string')


@Vows.batch
class PlacTests(Vows.Context):
    # These test are done in plac via the 'Interpreter' class, which
    # seems to be buggy at the moment...
    #def topic(self):
    #    import plac, interf
    #    return interf.generate_pics

    def shows_usage_message(self, topic):
        pass
        #expect(plac.call(topic, ['-h'])).to_be_like('')
        #expect(topic.check(' -h', 'cleared the shelve')).to_be_true()


class PointTests(Vows.Context):
    # TODO nest these tests, matching execution flow.
    def values_at_point_values(self, topic):
        error = 1e-12  # calculate suitable error based on w?
        val = interf.value_at_point(topic[0], topic[1][0], topic[1][1],
                zoom=4, time=10)  # vary zoom and time?
        expect(val).to_be_less_than(error)

    def gen_pmap_values(self, topic):
        # vary zoom and time?
        pmap = interf.gen_pmap(topic[0], zoom=4, time=10)
        val = tuple(pmap[topic[1][0], topic[1][1]])
        expect(val).to_equal((0, 0, 0))


def XRandomPoints(n):
    class XRandomPoints(PointTests):
        def topic(self, base):
            for p in range(n):
                yield (base, (randint(0, base.x - 1), randint(0, base.y - 1)))
    return XRandomPoints


@Vows.batch
class TotalCancellation(Vows.Context):
    '''Tests how classes behave when there are two points that cancel
    each other out.'''  # FIXME Is it alright to put a full-stop here?
    def topic(self):
        x = randint(60, 100)
        y = randint(60, 100)
        w = randint(1, 100)
        p = random()
        gx = random()
        gy = random()
        base = interf.Base(x, y, 100, 100, None, None)
        test_data = (
            (gx, gy, w, p),
            (gx, gy, w, p + 0.5), )
        for p in test_data:
            base.addPoint(*p)
        return base

#    class TwentyRandomPoints(XRandomPoints(20)): pass

    class SpecificPoints(PointTests):
        def topic(self, base):
            test_points = ((0, 0), (0, base.y - 1),
                    (base.x / 2, base.y / 2),
                    (base.x - 1, 0), (base.x - 1, base.y - 1), )
            for p in test_points:
                yield (base, p)


interf.py
#!/usr/bin/env python2.7

import os
from os.path import join as joinpath
import numpy
from datetime import datetime as dt
from PIL import Image
from math import pi, sqrt, cos
from collections import namedtuple
import plac
from clint.textui import progress
#from clint import args
#from clint.textui import puts, colored, indent

# inspired by http://lybniz2.sourceforge.net/safeeval.html
safe_funs = ['cos', 'cosh', 'e', 'exp', 'log', 'log10', 'pi', 'sin', 'sinh',
    'tan', 'tanh']
safe_funs = dict([(k, locals().get(k, None)) for k in safe_funs])
safe_funs['abs'] = abs


def value_at_point(base, x, y, zoom, time):
    # maybe move this elsewhere? like a 'HACKING' file (dev docs) or possibly
    # with the tests once corresponding tests have been created. # TODO
    '''Returns an integer value in [0,256) representing a pixel value fmod and
    zoom are used to 'zoom' in and out. Infact, what happens is the resolution
    stays the same, the points are just moved closer together and centered;
    then the frequency is modified so the relative distances are the same,
    e.g. If I use zoom = 2, the point co-ordinates are all halved, moving the
    points closer to each other and centered; then, The 'travel' is adapted
    accordingly to ensure the waves move slower across a (relatively) larger
    area (but with frequencies unchanged).

   01234567890123456789        01234567890123456789       01234567890123456789
   --------------------        --------------------       --------------------
 0|x(0,0)              |     0|x(0,0) .(0,7)       |    0|     |              |
 1|         (1,15).    |     1| .(1,1)             |    1|     |              |
 2|                    |     2|     .(2,5)         |    2|-----x(2,5) .(2,12) |
 3|   .(3,3)           | ->  3|                    | -> 3|      .(3,6)        |
 4|                    | ->  4|                    | -> 4|          .(4,10)   |
 5|           .(5,11)  |     5|                    |    5|                    |
 6|                    |     6|                    |    6|                    |
 7|                    |     7|                    |    7|                    |
   --------------------        --------------------       --------------------

     (1,15) <2,16>       ->        <1,8> (0,7)       ->         (2,12)
      (3,3) <4,4>        ->        <2,2> (1,1)       ->         (3,6)
     (5,11) <6,12>       ->        <3,6> (2,5)       ->         (4,10)
    '''
    # FIXME For some reason, I can see a spiral pattern in the darker patterns,
    # need to figure out why... the 'total cancellation' seem to rule out error
    # in that simple case, maybe the error of sin changes e.g. with frequency?
    global points
    contrib = []
    for p in base.points:
        distx = (2 * p.x + base.x * (zoom - 1)) / (zoom * 2)
        disty = (2 * p.y + base.y * (zoom - 1)) / (zoom * 2)
        # a^2 + b^2 = c^2
        dist = sqrt((x - distx) ** 2 + (y - disty) ** 2)  # in pixels
        dist = (zoom * dist) / base.travel  # in ms, adjusted for zoom
        totaltime = time + dist  # time at source+base.travel time=total time
        aphase = (totaltime * p.w) / 1000.0  # phase change during this time
        totalphase = p.p + aphase  # initial phase+addition phase=total phase
        val = cos(2 * pi * totalphase)
        contrib.append(val)
    return abs(sum(contrib))


def gen_pmap(base, zoom, time):
    '''Generates a numpy array representing the image
     Returns the numpy array.'''
    pmap = numpy.zeros((base.x, base.y, 3)).astype('uint8')
    scale = (255.0 / len(base.points))
    for x in range(base.x):
        for y in range(base.y):
            val = scale * value_at_point(base, x, y, zoom, time)
            pmap[x, y, 0] = val
            pmap[x, y, 1] = val
            pmap[x, y, 2] = val
    return pmap


# TODO calculate length for fixed strings of the form "%03d", or "%3d"
def generate_pics(base):
    timestamp = dt.now().strftime("%Y%m%d%H%M%S")  # TODO make this optional?
    frames = base.l / base.ft
    safe_funs['N'] = frames
    for n in progress.bar(range(1, frames + 1)):
        safe_funs['n'] = float(n)
        # potentially unsafe, see comment above 'safe_fun'
        zoom = eval(base.zs, {"__builtins__": None}, safe_funs)
        # DEBUG
        #print "%3d/%d" % (n, frames),
        #t = dt.now()
        pmap = gen_pmap(base, zoom, n * base.ft)
        # DEBUG
        #dura = (dt.now()-t).microseconds
        #print "%3d,%03dms" % (dura/1000, dura%1000)
        Image.fromarray(pmap).save(joinpath(base.tp, '%s-%03d.png') \
                % (timestamp, n))
    print '''Now use 'apngasm <filepath>/<filename>.png %s/%s*.png 1 %d' to
        create the animated png!''' % (base.tp, timestamp, base.resolution)


class Base():
    'class representing base values'

    def __init__(self, x, y, ft, l, tp, zs):
        self.x = x
        self.y = y
        self.ft = ft
        self.l = l
        self.tp = tp
        self.zs = zs
        self.points = []

    class Point(namedtuple('Point', 'base, gx gy w p')):
        '''class representing point source in image.
        gx,gy are co-ordinates of point,
        w is wavelength in revolutions per second,
        p is initial phase of point in turns'''
        __slots__ = ()

        @property
        def x(self):
            return int(self.gx * self.base.x)

        @property
        def y(self):
            return int(self.gy * self.base.y)

    @property
    def travel(self):
        'pixel distance covered in a millisecond'
        return self.y / 1000.0

    @property
    def resolution(self):
        '''value for x in 1/x representing frametime as a fraction of a second,
        for use as an argument to apngasm'''
        return 1000 / self.ft

    def addPoint(self, gx, gy, w, p):
        self.points.append(self.Point(self, gx, gy, w, p))


@plac.annotations(
    x_size=('height of image, in pixels', 'option', 'x', int),
    y_size=('width of image in pixels', 'option', 'y', int),
    tmp_path=('temporary path to save files in', 'option'),
    zoom_string=('''
    Python expression representing how the animation should zoom, use 'n' to
    represent the image number, and 'N' for the total number of frames
    ''', 'option', 'z', None, None, 'ZOOM_STRING'),
    length=('''
    How long the animation should be, in milliseconds (relative to base.ft)
    ''', 'option', 'l', int),
)  # Can I make y_size depend on x_size?
def main(x_size=60, y_size=120, tmp_path=os.path.expanduser('~/intpics'),
        zoom_string='4', length=2000):
    "Generate a series of png files from which to create an apng file."
    if not os.path.exists(tmp_path):
        os.makedirs(tmp_path)
    base = Base(y_size, x_size, 100, length, tmp_path, zoom_string)
    # TODO point are fixed, add way to specify them - maybe from csv format?
    base.addPoint(1.0 / 2, 1.0 / 8, 1.0 / 4, 0.75)
    base.addPoint(1.0 / 3, 3.0 / 5, 1, 0.5)
    base.addPoint(1.0 / 8, 2.0 / 5, 1.0 / 3, 1)
    generate_pics(base)

if __name__ == '__main__':
    plac.call(main)
